### ISSUE TITLE
Applying SOL_SOCKET/SO_ATTACH_REUSEPORT_CBPF option to Http/3 UDP listener failed

### ISSUE THREAD

**Issue by AMagnolia** [envoyproxy/envoy#36005]
*Title*: *Applying SOL_SOCKET/SO_ATTACH_REUSEPORT_CBPF option to Http/3 UDP listener failed*

*Description*:
>I am trying to test the configuration provided in https://github.com/envoyproxy/envoy/blob/a4058c3751c67b4a42b909dcab8e358952ed03ca/configs/envoyproxy_io_proxy_http3_downstream.yaml. The client side proxy configuration is as below:
 ```
admin:
  address:
    socket_address:
      protocol: TCP
      address: 0.0.0.0
      port_value: 9901
static_resources:
  listeners:
  - name: listener_tcp
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 9879
    filter_chains:
    - transport_socket:
        name: envoy.transport_sockets.tls
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
          common_tls_context:
            tls_certificates:
            - certificate_chain:
                  filename: /etc/envoy/test.cer
               private_key:
                  filename: /etc/envoy/testserver.key
      filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          codec_type: HTTP2
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              response_headers_to_add:
              - header:
                  key: alt-svc
                  value: h3=":9879"; ma=86400, h3-29=":9879"; ma=86400
              domains: ["*"]
              routes:
              - match:
                  prefix: "/"
                route:
                  host_rewrite_literal: www.envoyproxy.io
                  cluster: service_envoyproxy_io
          http3_protocol_options:
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router

    - name: listener_udp
      address:
        socket_address:
          protocol: UDP
          address: 0.0.0.0
          port_value: 9879
      reuse_port: true
      udp_listener_config:
        quic_options: {}
        downstream_socket_config:
      filter_chains:
      - transport_socket:
          name: envoy.transport_sockets.quic
          typed_config:
            '@type': type.googleapis.com/envoy.extensions.transport_sockets.quic.v3.QuicDownstreamTransport
            downstream_tls_context:
              common_tls_context:
                tls_certificates:
                - certificate_chain:
                    filename: /etc/envoy/test.cer
                  private_key:
                    filename: /etc/envoy/testserver.key
      filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          codec_type: HTTP3
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/"
                route:
                  host_rewrite_literal: www.envoyproxy.io
                  cluster: service_envoyproxy_io
          http3_protocol_options:
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
  clusters:
  - name: service_envoyproxy_io
    connect_timeout: 30s
    type: LOGICAL_DNS
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: service_envoyproxy_io
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: www.envoyproxy.io
                port_value: 443
    transport_socket:
      name: envoy.transport_sockets.tls
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
        sni: www.envoyproxy.io
 ```
And the common to run the docker is:
`/usr/bin/docker run -idt --name envoy_TEST -e ENVOY_UID=0 -v "/root/envoyTest/envoy.yaml:/etc/envoy/envoy.yaml" -v "/root/envoyTest:/tmp" -v "/root/envoyTest/tracepre.pb:/etc/envoy/trace.pb" -v "/root/envoy/trade.cer:/etc/envoy/trade.cer" -v "/root/envoy/test.cer:/etc/envoy/test.cer" -v "/root/envoy/testserver.key:/etc/envoy/testserver.key" -v "/root/envoy/server.key:/etc/envoy/server.key" -v "/root/envoy/Test.cer:/etc/envoy/Test.cer" -v "/root/envoy/Testserver.key:/etc/envoy/Testserver.key" -p 9879:9879 envoyproxy/envoy:v1.18.2 #-l debug -c /etc/envoy/envoy.yaml`

*Repro steps*:
> the linux kernel version is 3.10.0 and the kernel headers package has been installed. Executing 
`sudo find /usr/include/ -name bpf.h` , 
the result is: 
`/usr/include/linux/bpf.h`

*It fails. The output is*:
>[warning][connection] [source/common/network/socket_option_impl.cc:27] Setting SOL_SOCKET/SO_ATTACH_REUSEPORT_CBPF option on socket failed: Protocol not available
[warning][quic] [source/common/quic/active_quic_listener.cc:59] Failed to apply socket options to socket 49 on listener listener_udp after binding
[error][main] [source/server/worker_impl.cc:55] failed to add listener on worker: Failed to apply socket options.
[warning][main] [source/server/server.cc:642] there is no configured limit to the number of allowed active connections. Set a limit via the runtime key overload.global_downstream_max_connections
[error][config] [source/server/listener_manager_impl.cc:719] listener 'listener_udp' failed to listen on address '0.0.0.0:9879' on worker

Thank you.

**Comment by phlax** [comment]
cc @mattklein123 @danzh2010

**Comment by danzh2010** [comment]
From https://man7.org/linux/man-pages/man7/socket.7.html:
```UDP support for this feature is available since Linux 4.5;
TCP support is available since Linux 4.6.```
